File: solutions/0451. Sort Characters By Frequency/0451.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  string frequencySort(string_view s) {
    const int n = s.length();
    string ans;
    vector<int> count(128);
    vector<vector<char>> buckets(n + 1);

    for (const char c : s)
      ++count[c];

    for (int i = 0; i < 128; ++i) {
      const int freq = count[i];
      if (freq > 0)
        buckets[freq].push_back((char)i);
    }

    for (int freq = n; freq > 0; --freq)
      for (const char c : buckets[freq])
        ans += string(freq, c);

    return ans;        
  }
};

File: solutions/0451. Sort Characters By Frequency/0451.rs
// Time complexity: O(n)
// Space complexity: O(n)
use std::collections::VecDeque;

impl Solution {
    pub fn frequency_sort(s: String) -> String {
        let n = s.len();
        let mut ans = String::new();
        let mut count = vec![0; 128];
        let mut buckets = vec![VecDeque::new(); n + 1];

        for c in s.chars() {
            count[c as usize] += 1;
        }

        for i in 0..128 {
            let freq = count[i];
            if freq > 0 {
                buckets[freq].push_back(i as u8 as char);
            }
        }

        for freq in (1..=n).rev() {
            for c in buckets[freq].iter() {
                ans += &c.to_string().repeat(freq);
            }
        }

        ans
    }
}
